src/jit/codegenxarch.cpp

```diff
@@ -7356,6 +7356,142 @@ void CodeGen::genSSE2BitwiseOp(GenTreePtr treeNode)
     inst_RV_RV(ins, targetReg, operandReg, targetType);
 }
 
+//-----------------------------------------------------------------------------------------
+// genSSE41RoundOp - generate SSE41 code for the given tree as a round operation
+//
+// Arguments:
+//    treeNode  - tree node
+//
+// Return value:
+//    None
+//
+// Assumptions:
+//     i) SSE4.1 is supported by the underlying hardware
+//    ii) tree oper is a GT_INTRINSIC
+//   iii) tree type is floating point type
+//    iv) tree oper is CORINFO_INTRINSIC_Round, _Ceiling, or _Floor
+//     v) caller of this routine needs to call genProduceReg()
+void CodeGen::genSSE41RoundOp(GenTreeOp* treeNode)
+{
+    // i) SSE4.1 is supported by the underlying hardware
+    assert(compiler->compSupports(InstructionSet_SSE41));
+
+    // ii) tree oper is a GT_INTRINSIC
+    assert(treeNode->OperGet() == GT_INTRINSIC);
+
+    GenTree* srcNode = treeNode->gtGetOp1();
+
+    // iii) tree type is floating point type
+    assert(varTypeIsFloating(srcNode));
+    assert(srcNode->TypeGet() == treeNode->TypeGet());
+
+    genConsumeOperands(treeNode);
+
+    instruction ins  = (treeNode->TypeGet() == TYP_FLOAT) ? INS_roundss : INS_roundsd;
+    emitAttr    size = emitTypeSize(treeNode);
+
+    regNumber dstReg = treeNode->gtRegNum;
+    regNumber srcReg = srcNode->gtRegNum;
+
+    unsigned ival = 0;
+
+    // iv) tree oper is CORINFO_INTRINSIC_Round, _Ceiling, or _Floor
+    switch (treeNode->gtIntrinsic.gtIntrinsicId)
+    {
+        case CORINFO_INTRINSIC_Round:
+            ival = 4;
+            break;
+
+        case CORINFO_INTRINSIC_Ceiling:
+            ival = 10;
+            break;
+
+        case CORINFO_INTRINSIC_Floor:
+            ival = 9;
+            break;
+
+        default:
+            ins = INS_invalid;
+            assert(!"genSSE41: unsupported oper");
+            unreached();
+            break;
+    }
+
+    if (srcNode->isContained())
```

Comment:
Just as a side note. I mostly modeled this after the code paths in `emitInsBinary` and noticed that the current code (in `emitInsBinary`) ends up duplicating several checks multiple times. For example, it calls `isContainedIndir`, then `isContainedIntFltOrDblImmed`, then `isLclFldUsedFromMemory`, all of which repeat the `isContained` check.

Reply:
I tried to eliminate those duplicate checks here and would imagine `emitInsBinary` would benefit from similar "cleanup"

Reply:
`emitInsBinary` is a bloody mess. And it doesn't belong in the emitter, it should be in CodeGen. I actually refactored much of that stuff a while ago but then I found more interesting things to do and didn't finish it.